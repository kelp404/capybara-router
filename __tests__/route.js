const React = require('react');
const Route = require('../lib/route');

test('Initial Route without a parent.', () => {
  const route = new Route({
    name: 'web',
    uri: '/users/{userId:[\\w-]{20}}/projects?index',
    resolve: {
      user: ({userId}) => ({id: userId, name: 'User'}),
      projects: () => ([
        {id: 'AWgrmJp1SjjuUM2bzZXM', title: 'Project'},
      ]),
    },
    component: () => <div/>,
  });
  expect(route).toMatchSnapshot();
});

test('Initial Route with a parent.', () => {
  const parent = new Route({name: 'web', uri: '/'});
  const route = new Route({
    name: 'web.dashboard',
    uri: 'dashboard',
    parent,
  });
  expect(route).toMatchSnapshot();
});

test('Generate a route with the parent.', () => {
  const parent = new Route({
    name: 'web',
    uri: '/users/{userId:[\\w-]{20}}/projects?index?sort',
    resolve: {id: () => 'id'},
    onEnter() {},
    component() {},
  });
  const child = new Route({
    name: 'web.project',
    uri: '/users/{userId:[\\w-]{20}}/projects/{projectId:[\\w-]{20}}',
    resolve: {id: () => 'id'},
    onEnter() {},
    component() {},
    parent,
  });
  expect(parent).toMatchSnapshot();
  expect(child).toMatchSnapshot();
});

test('Get an error on generating a route with a resolve key called "key".', () => {
  const func = () => new Route({
    name: 'web',
    uri: '/',
    resolve: {key: () => null},
  });

  expect(func).toThrow(Error);
});
test('Get an error on generating a route with a resolve key called "params".', () => {
  const func = () => new Route({
    name: 'web',
    uri: '/',
    resolve: {params: () => null},
  });

  expect(func).toThrow(Error);
});

test('Generate the URI of the route with params', () => {
  const fakeRoute = new Route({
    name: 'web',
    uri: '/users/{userId:[\\w-]{20}}/projects?index?sort?nullValue?empty?array?undefinedValue',
    resolve: {
      user: ({userId}) => Promise.resolve({id: userId, name: 'User'}),
      projects: () => Promise.resolve([{id: 'AWgrmJp1SjjuUM2bzZXM', title: 'Project'}]),
    },
    component: () => <div/>,
  });
  const uri = fakeRoute.generateUri({
    userId: 'AWgrmJp1SjjuUM2bzZXM',
    index: 0,
    sort: 'asc',
    nullValue: null,
    empty: '',
    undefinedValue: undefined,
    array: ['a', 'b'],
  });
  expect(uri).toMatchSnapshot();
});
