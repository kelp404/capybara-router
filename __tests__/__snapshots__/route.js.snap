// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generate a route with the parent. 1`] = `
Route {
  "component": [Function],
  "dismissalDelay": undefined,
  "generateLocation": [Function],
  "generateUri": [Function],
  "isAbstract": false,
  "loadComponent": undefined,
  "matchPattern": "/users/([\\\\w-]{20})/projects",
  "matchReg": /\\^\\\\/users\\\\/\\(\\[\\\\w-\\]\\{20\\}\\)\\\\/projects\\$/,
  "name": "web",
  "onEnter": [Function],
  "parents": Array [],
  "resolve": Object {
    "id": [Function],
  },
  "uri": "/users/{userId:[\\\\w-]{20}}/projects?index?sort",
  "uriParamKeys": Array [
    "userId",
    "?index",
    "?sort",
  ],
  "uriTemplate": "/users/{userId}/projects",
}
`;

exports[`Generate a route with the parent. 2`] = `
Route {
  "component": [Function],
  "dismissalDelay": undefined,
  "generateLocation": [Function],
  "generateUri": [Function],
  "isAbstract": false,
  "loadComponent": undefined,
  "matchPattern": "/users/([\\\\w-]{20})/projects/users/([\\\\w-]{20})/projects/([\\\\w-]{20})",
  "matchReg": /\\^\\\\/users\\\\/\\(\\[\\\\w-\\]\\{20\\}\\)\\\\/projects\\\\/users\\\\/\\(\\[\\\\w-\\]\\{20\\}\\)\\\\/projects\\\\/\\(\\[\\\\w-\\]\\{20\\}\\)\\$/,
  "name": "web.project",
  "onEnter": [Function],
  "parents": Array [
    Route {
      "component": [Function],
      "dismissalDelay": undefined,
      "generateLocation": [Function],
      "generateUri": [Function],
      "isAbstract": false,
      "loadComponent": undefined,
      "matchPattern": "/users/([\\\\w-]{20})/projects",
      "matchReg": /\\^\\\\/users\\\\/\\(\\[\\\\w-\\]\\{20\\}\\)\\\\/projects\\$/,
      "name": "web",
      "onEnter": [Function],
      "parents": Array [],
      "resolve": Object {
        "id": [Function],
      },
      "uri": "/users/{userId:[\\\\w-]{20}}/projects?index?sort",
      "uriParamKeys": Array [
        "userId",
        "?index",
        "?sort",
      ],
      "uriTemplate": "/users/{userId}/projects",
    },
  ],
  "resolve": Object {
    "id": [Function],
  },
  "uri": "/users/{userId:[\\\\w-]{20}}/projects/{projectId:[\\\\w-]{20}}",
  "uriParamKeys": Array [
    "userId",
    "?index",
    "?sort",
    "userId",
    "projectId",
  ],
  "uriTemplate": "/users/{userId}/projects/users/{userId}/projects/{projectId}",
}
`;

exports[`Generate the URI of the route with params 1`] = `"/users/AWgrmJp1SjjuUM2bzZXM/projects?array=a&array=b&empty=&index=0&nullValue&sort=asc"`;

exports[`Initial Route with a parent. 1`] = `
Route {
  "component": undefined,
  "dismissalDelay": undefined,
  "generateLocation": [Function],
  "generateUri": [Function],
  "isAbstract": false,
  "loadComponent": undefined,
  "matchPattern": "/dashboard",
  "matchReg": /\\^\\\\/dashboard\\$/,
  "name": "web.dashboard",
  "onEnter": undefined,
  "parents": Array [
    Route {
      "component": undefined,
      "dismissalDelay": undefined,
      "generateLocation": [Function],
      "generateUri": [Function],
      "isAbstract": false,
      "loadComponent": undefined,
      "matchPattern": "/",
      "matchReg": /\\^\\\\/\\$/,
      "name": "web",
      "onEnter": undefined,
      "parents": Array [],
      "resolve": Object {},
      "uri": "/",
      "uriParamKeys": Array [],
      "uriTemplate": "/",
    },
  ],
  "resolve": Object {},
  "uri": "dashboard",
  "uriParamKeys": Array [],
  "uriTemplate": "/dashboard",
}
`;

exports[`Initial Route without a parent. 1`] = `
Route {
  "component": [Function],
  "dismissalDelay": undefined,
  "generateLocation": [Function],
  "generateUri": [Function],
  "isAbstract": false,
  "loadComponent": undefined,
  "matchPattern": "/users/([\\\\w-]{20})/projects",
  "matchReg": /\\^\\\\/users\\\\/\\(\\[\\\\w-\\]\\{20\\}\\)\\\\/projects\\$/,
  "name": "web",
  "onEnter": undefined,
  "parents": Array [],
  "resolve": Object {
    "projects": [Function],
    "user": [Function],
  },
  "uri": "/users/{userId:[\\\\w-]{20}}/projects?index",
  "uriParamKeys": Array [
    "userId",
    "?index",
  ],
  "uriTemplate": "/users/{userId}/projects",
}
`;
